tests/foreman/cli/test_auth.py

```diff
@@ -0,0 +1,441 @@
+        echo '  :use_sessions: {1}' >> {0}'''.format(
+            HAMMER_CONFIG,
+            'true' if enable else 'false'
+        )
+    )
+    if (result.return_code == 0 and add_default_creds):
+        result = ssh.command(
+            '''{{ echo '  :username: "{0}"';\
+            echo '  :password: "{1}"'; }} >> {2}'''.format(
+                settings.server.admin_username,
+                settings.server.admin_password,
+                HAMMER_CONFIG
+                )
+            )
+    return result.return_code
+
+
+@run_in_one_thread
+class HammerAuthTestCase(CLITestCase):
+    """Implements hammer authentication tests in CLI"""
+
+    @classmethod
+    def setUpClass(cls):
+        """Creates users to be reused across tests"""
+        super(HammerAuthTestCase, cls).setUpClass()
+        cls.uname_admin = gen_string('alpha')
+        cls.uname_viewer = gen_string('alpha')
+        cls.password = gen_string('alpha')
+        cls.mail = 'test@example.com'
+        make_user({
+                'login': cls.uname_admin,
+                'password': cls.password,
+                'admin': '1',
+        })
+        make_user({
+                'login': cls.uname_viewer,
+                'password': cls.password,
+        })
+        User.add_role({'login': cls.uname_viewer, 'role': 'Viewer'})
+
+    @classmethod
+    def tearDownClass(cls):
+        """Making sure sessions are disabled after test run"""
+        configure_sessions(False)
+
+    @tier1
+    def test_positive_create_session(self):
+        """Check if user stays authenticated with session enabled
+
+        :id: fcee7f5f-1040-41a9-bf17-6d0c24a93e22
+
+        :Steps:
+
+            1. Set use_sessions, set short expiration time
+            2. Authenticate, assert credentials are not demanded
+               on next command run
+            3. Wait until session expires, assert credentials
+               are required
+
+        :expectedresults: The session is successfully created and
+            expires after specified time
+        """
+        try:
+            idle = Settings.list({'search': 'name=idle_timeout'})[0][u'value']
+            Settings.set({'name': 'idle_timeout', 'value': 1})
+            result = configure_sessions()
+            self.assertEqual(result, 0, 'Failed to configure hammer sessions')
+            Auth.login({
+                'username': self.uname_admin,
+                'password': self.password
+            })
+            result = Auth.status({
+                 'hammer_options': {'username': u'', 'password': u''}})
+            self.assertIn(
+                LOGEDIN_MSG.format(self.uname_admin),
+                result[0][u'message']
+            )
+            # list organizations without supplying credentials
+            with self.assertNotRaises(CLIReturnCodeError):
+                Org.list({
+                    'hammer_options': {'username': u'', 'password': u''}})
```

Comment:
Ok, this approach isn't cool either, but it gave me another idea which is hopefully more or less acceptable.
My personal list of why it's wrong:
1) Usage from user perspective. If user don't want to pass credentials, he's supposed to pass extra dict with empty username and password as an extra key. In the end every single command looks like `Org.list({'hammer_options': {'username': u'', 'password': u''}})` which is very long and quite ugly format.
2) We're breaking the idea behind `options` dict. Before it was simple - that was a dict of key/value pairs which are passed as command parameters. Now it has an exclusion, which acts differently, which brings extra logic and btw is not even documented 🙂 
3) Lots of core changes. Command is constructed and executed quite differently, lots of changes are breaking changes - even `command` arg isn't a string with command anymore but a list of 2 strings. We should be very careful introducing such changes as they can easily introduce new regressions and personally i'd like to avoid such changes as much as possible.

So, extra arg sounds like a bad idea, extra key in existing dict isn't much better too, what's remaining? :) Your format reminded me our context manager which basically injects in class and passes right credentials. It's `with_user()` CM, it usage is smth like this:
```python
Org.with_user(username, password).list()
```
`Base.execute()` uses creds provided as arguments, if not found - searches for them as class attributes, if not found - uses the ones provided in properties file. The idea behind `with_user()` is to set class attributes. Why not to extend its logic then?
The idea off the top of my head without deep analysis - different `with_user()` logic.
1) If username/password are `None` - set class attributes as some empty strings. This way class attributes are set, so `Base.execute()` will use them (empty strings) and won't load the ones from properties file. 
2) Secondly, very small `Base.execute()` change - instead of blindly passing username and password, pass them with '-u' and '-p' correspondingly only in case they're not empty (not empty strings).
2a) If 'interactive' flag is needed (although i can't see any usage of it in this PR) - it can be implemented similar way as an extra class parameter introduced by `with_user()` CM.
3) Usage. In the end every command which should be executed without providing any creds will look like this:
```python
Org.with_user().list()
```
Yeah, not great, but definitely shorter, and, basically, pretty understandable approach - when we wanna pass different creds we call it like `Org.with_user(username, password).list()`, so when we don't want to pass any creds - we use similar approach but don't specify any creds at all, simple as that 🙂 

What do you think? Imo it's still not perfect solution, but definitely better because of less breaking changes and better usability.
Feel free to propose your ideas

Reply:
Make sense, wish I knew about with_user sooner, one always learns :) About the  "--interactive no" setting, atm it is needed just for the case when we don't provide creds and session is not enabled, in such case we want hammer to fail and not prompt for creds. Will check in soon with updated code, cheers